### PR TITLE
sd-journal: several fixes for generic_array_bisect()

### LINKED ISSUE #27533: Failed assertion when parsing corrupted journal
### systemd version the issue has been seen with

latest main

### Used distribution

Fedora Rawhide

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

systemd-journald

### Expected behaviour you didn't see

_No response_

### Unexpected behaviour you saw

It's that time of the day again:

```
$ build-san/journalctl --file /tmp/id\:000000\,sig\:06\,src\:000711\,time\:110015157\,execs\:33104794\,op\:MOpt_havoc\,rep\:2                                                                                                                  
Assertion 'last_index <= right' failed at src/libsystemd/sd-journal/journal-file.c:2873, function generic_array_bisect(). Aborting.                                                                                                            
Aborted (core dumped)  
```

Journal: https://mrc0mmand.fedorapeople.org/id:000000,sig:06,src:000711,time:110015157,execs:33104794,op:MOpt_havoc,rep:2.zst
(ready to be dropped in `test/test-journals/corrupted/` together with a possible fix)

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #27533.

### CODE REVIEW

File: src/libsystemd/sd-journal/journal-file.c [hunk 1]
```diff
@@ -2775,6 +2775,59 @@ enum {
...
+
+        p = journal_file_entry_array_item(f, array, i);
+        if (p <= 0)
+                r = -EBADMSG;
+        else
+                r = test_object(f, p, needle);
+        if (IN_SET(r, -EBADMSG, -EADDRNOTAVAIL)) {
+                log_debug_errno(r, "Encountered invalid entry while bisecting, cutting algorithm short.");
+                *right = i;
+                return -ENOANO; /* recognizable error */
```
Comment (mrc0mmand) [inline src/libsystemd/sd-journal/journal-file.c:2812]: And here I thought that nothing would beat `ENOTNAM` when it comes to cryptic errnos, nice :)

**Review by DaanDeMeyer — APPROVED** [review]

**Comment by mrc0mmand** [comment]
Thank you! I gave this a spin with the ~7.5k of corrupted journals I currently have at hand and nothing crashed; sanitizers were happy as well.

**Comment by vcaputo** [comment]
> Thank you! I gave this a spin with the ~7.5k of corrupted journals I currently have at hand and nothing crashed; sanitizers were happy as well.

Just out of curiosity (and maybe for any readers): would your test know if these changes just prevented the operation from doing the work?  I assume yes, but it'd be nice to see confirmation that this can't possibly be 7.5k of corrupted journals NOOP'd into success now.

**Comment by evverx** [comment]
@vcaputo this test makes sure that stuff doesn't crash. It can't confirm that the code works correctly.

**Comment by mrc0mmand** [comment]
> > Thank you! I gave this a spin with the ~7.5k of corrupted journals I currently have at hand and nothing crashed; sanitizers were happy as well.
> 
> Just out of curiosity (and maybe for any readers): would your test know if these changes just prevented the operation from doing the work? I assume yes, but it'd be nice to see confirmation that this can't possibly be 7.5k of corrupted journals NOOP'd into success now.

Yeah, as Evgeny said, it just makes sure `journalctl` doesn't die with a signal, it doesn't check if the action taken makes "sense", since there's just way too many error paths. But for that we have the integration test suite, which should do these checks instead (to some degree).

**Comment by evverx** [comment]
@mrc0mmand on a somewhat related note I think it would be great to run `journalctl` under Valgrind with that corpus. It should help to somewhat replace MSan (which would be used on OSS-Fuzz if that AFL++ PR got merged). That would be slow though so it would make sense only if it could be incorporated in a daily/weekly cron job.

**Comment by mrc0mmand** [comment]
> @mrc0mmand on a somewhat related note I think it would be great to run `journalctl` under Valgrind with that corpus. It should help to somewhat replace MSan (which would be used on OSS-Fuzz if that AFL++ PR got merged).

That sounds like a good idea, especially given your recent findings :)

> That would be slow though so it would make sense only if it could be incorporated in a daily/weekly cron job.

Depends. Running it only once against the directory full of journals (e.g. `journalctl --directory=...)` should be fairly quick and could be incorporated directly into TEST-04-JOURNAL. However, running it for every journal would indeed be quite slow (it's slow even without valgrind).

**Pull request merged**